### ISSUE TITLE
[Feature Request] Calculation or Formula field

### ISSUE THREAD

**Issue by LemarinelNet** [Laravel-Backpack/CRUD#3099]
# Feature Request

### What's the feature you think Backpack should have?

I've been tinkering with backpack for a few hours, and haven't found any official way to do it : 
I'd like to have a field that display a value computed from two other fields.
The use case would be this one : 
On a "item" CRUD, I have those fields : 
- purchase price
- margin
- [ a calculated value, which is the "minimum sell price"] : purchase price * margin
- sell price
- [ a calculated value, which is the "gross margin"] : sell price / purchase price * 100

Those two fields do have to be refresh every time the user change any of the two values.

### Have you already implemented a prototype solution, for your own project?

Not yet. I can barely imagine something like this : 
- Need for a javascript helper function, that would return the value of any input field . ex : $.CRUD("purchase_price")
- The calculated field would be defined as this : 
```
                'name' => 'minimum_sell_price',
                'type' => 'computed',
                'label' => "min. sell price.",
                'depends_on' => ["purchase_price","margin"],
                'formula' => " $.CRUD('purchase_price') * $.CRUD('margin') ",
                // optionnal settings
                'suffix' => " €",
```
- When such a field is used, it binds an event on each "depends_on" fields.
- When the event triggers, it evaluates the formula and sets the field value.
- The formula would not be limited to maths, it could be any JS processing the dev would like to hack ;)... Ajax call to retrieve an image, some other text, etc....

### Do you see this as a core feature or an add-on?

Core, I see many use cases ;)

**Comment by tabacitu** [comment]
> I see many use cases ;)

I agree with you on that one @LemarinelNet ! 😆 In fact, I see so many use cases that it's difficult to create something general enough to cover all use cases, I think. It's one of the reasons why we haven't created anything like this yet. But there are other problems with this that... don't feel quite right. I'll explain in more detail below.

---

Your use case, though, where you just add a formula to the field and it calculates that formula using JS, shouldn't be too difficult to achieve by [creating a custom field type](https://backpackforlaravel.com/docs/4.1/crud-fields#creating-a-custom-field-type). You could start from the `number` field:

```diff
-<!-- number input -->
+<!-- total input -->
@include('crud::fields.inc.wrapper_start')
    <label>{!! $field['label'] !!}</label>
    @include('crud::fields.inc.translatable_icon')

    @if(isset($field['prefix']) || isset($field['suffix'])) <div class="input-group"> @endif
        @if(isset($field['prefix'])) <div class="input-group-prepend"><span class="input-group-text">{!! $field['prefix'] !!}</span></div> @endif
        <input
        	type="number"
        	name="{{ $field['name'] }}"
+               data-init-function="bpFieldInitTotal"
+               data-formula="{{ $field['formula'] }}"
            value="{{ old(square_brackets_to_dots($field['name'])) ?? $field['value'] ?? $field['default'] ?? '' }}"
            @include('crud::fields.inc.attributes')
        	>
        @if(isset($field['suffix'])) <div class="input-group-append"><span class="input-group-text">{!! $field['suffix'] !!}</span></div> @endif

    @if(isset($field['prefix']) || isset($field['suffix'])) </div> @endif

    {{-- HINT --}}
    @if (isset($field['hint']))
        <p class="help-block">{!! $field['hint'] !!}</p>
    @endif
@include('crud::fields.inc.wrapper_end')
```

and at the bottom of the file add the JS logic it needs, something like:

```php

{{-- ########################################## --}}
{{-- Extra CSS and JS for this particular field --}}
{{-- If a field type is shown multiple times on a form, the CSS and JS will only be loaded once --}}
@if ($crud->fieldTypeNotLoaded($field))
    @php
        $crud->markFieldTypeAsLoaded($field);
    @endphp

    {{-- FIELD JS - will be loaded in the after_scripts section --}}
    @push('crud_fields_scripts')
        <script>
            function bpFieldInitTotal(element) {
                // element will be a jQuery wrapped DOM node - the number input shown above
                element.val(eval(element.formula));
            }
        </script>
    @endpush

@endif
{{-- End of Extra CSS and JS --}}
{{-- ########################################## --}}
```

In the example above, I've done some changes to your addField syntax:

```diff
                'name' => 'minimum_sell_price',
-                'type' => 'computed',
+                'type' => 'total',
                'label' => "min. sell price.",
-                'depends_on' => ["purchase_price","margin"],
-                'formula' => " $.CRUD('purchase_price') * $.CRUD('margin') ",
+                'formula' => " $('input[name=purchase_price]').val() * $('input[name=margin]').val()",
                // optional settings
                'suffix' => " €",
```

`depends_on` is not really needed I think. Also `formula` should probably be JS code, complete with selectors and everything, so it's as broad as possible.

There are some caveats to this approach though:
- there can be no double-quotes inside the formula code (or they should be escaped or smth);
- you're basically writing JS and jQuery inside the PHP controller; I expect a lot of people won't like this 😄 
- the JS formula is run using `eval()`, which has [some big security implications](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#Never_use_eval!); though for admin panel usage, where you expect your admin to be trustworthy, that shouldn't be such a big problem, I think;
- you pretty much have to know the HTML output of the fields, to create the right selectors - yes most of the time it will be an input, but sometimes it'll be a textarea or something;

The field above should work pretty well, I think (haven't tested the code), but... I don't know if it should make it into the Backpack core, because of the problems above... I'll think about it, but in the meantime I'll move this issue to the add-ons repo, because I do think this would make for a very cool Backpack add-on; not everybody will need it, but those who will, will super-appreciate that it exists. Please let me know if you'll consider [turning this into an add-on](https://backpackforlaravel.com/docs/4.1/add-ons-tutorial) @LemarinelNet - I can help with feedback and stuff. If not, we _might_ consider creating an official add-on of it, but only if it garners some more support from the community, if more people ask for it.

Hope my directions above help. Please post here your final code if you end up with something that works well enough for you. And let us know if you'll be creating the add-on.

Cheers!
